File: typescript/src/normalize.ts
/**
 * Response normalization helpers.
 *
 * The Kairos microservice workers return resource-specific response shapes
 * (e.g. { tasks: [...], count, total, hasMore }) rather than the generic
 * { data: [...], pagination: {...} } envelope the SDK exposes to callers.
 *
 * These helpers bridge the gap without coupling every resource method to the
 * specific field names used by each worker.
 */

import type { PaginatedResponse } from './types';

/**
 * Normalise a paginated list response from a worker into the generic
 * PaginatedResponse<T> shape that the SDK surface exposes.
 *
 * @param raw    Raw JSON object returned by the worker.
 * @param key    The resource-specific array key (e.g. "tasks", "goals").
 * @param limit  The requested page size (used to synthesise `page` number).
 * @param offset The requested offset.
 */
export function normalizePaginated<T>(
  raw: Record<string, unknown>,
  key: string,
  limit = 20,
  offset = 0,
): PaginatedResponse<T> {
  // Worker may use a resource-specific key ("tasks") or the generic "data" key.
  const items = ((raw[key] ?? raw['data'] ?? []) as T[]);

  // Workers use "total" or "total_count" for the grand total.
  const total = Number(raw['total'] ?? raw['total_count'] ?? raw['count'] ?? 0);

  // Workers use "hasMore" (camelCase) or "has_more" (snake_case).
  const hasMore = Boolean(raw['hasMore'] ?? raw['has_more'] ?? false);

  // Actual limit returned by the worker (may differ from requested).
  const resolvedLimit = Number(raw['limit'] ?? limit);
  const resolvedOffset = Number(raw['offset'] ?? offset);

  return {
    data: items,
    pagination: {
      page: resolvedLimit > 0 ? Math.floor(resolvedOffset / resolvedLimit) + 1 : 1,
      limit: resolvedLimit,
      total,
      has_more: hasMore,
    },
  };
}

/**
 * Normalise a single-resource response from a worker.
 *
 * @param raw Raw JSON object returned by the worker.
 * @param key The resource-specific key (e.g. "task", "goal", "document").
 */
export function normalizeSingle<T>(raw: Record<string, unknown>, key: string): T {
  return (raw[key] ?? raw['data']) as T;
}

/**
 * Normalise a list sub-resource (e.g. assignees, labels, dependencies)
 * where the worker returns a plain array or wraps it in a keyed object.
 */
export function normalizeList<T>(raw: unknown, key?: string): T[] {
  if (Array.isArray(raw)) return raw as T[];
  if (raw && typeof raw === 'object') {
    const obj = raw as Record<string, unknown>;
    if (key && Array.isArray(obj[key])) return obj[key] as T[];
    if (Array.isArray(obj['data'])) return obj['data'] as T[];
  }
  return [];
}

File: python/kairos/_normalize.py
"""Response normalization helpers for the Kairos SDK.

The Kairos microservice workers return resource-specific response shapes
(e.g. ``{"tasks": [...], "count": N, "total": N, "hasMore": bool}``) rather
than the generic ``{"data": [...], "pagination": {...}}`` envelope the SDK
exposes to callers.  These helpers bridge the gap.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Type, TypeVar

T = TypeVar("T")


def normalize_paginated(
    raw: Dict[str, Any],
    key: str,
    model: Type[T],
    limit: int = 20,
    offset: int = 0,
) -> Dict[str, Any]:
    """Return a normalized paginated dict matching PaginatedResponse schema.

    Args:
        raw:    Raw JSON dict returned by the worker.
        key:    Resource-specific array key (e.g. "tasks", "goals").
        model:  Pydantic model class to instantiate each item.
        limit:  Requested page size (used to compute page number).
        offset: Requested offset.

    Returns a dict suitable for ``PaginatedResponse[T](**result)``.
    """
    items_raw: List[Any] = raw.get(key) or raw.get("data") or []
    items = [model(**item) for item in items_raw]

    total = int(raw.get("total") or raw.get("total_count") or raw.get("count") or 0)
    resolved_limit = int(raw.get("limit") or limit)
    resolved_offset = int(raw.get("offset") or offset)
    page = (resolved_offset // resolved_limit + 1) if resolved_limit > 0 else 1
    # Some workers (e.g. documents, whiteboards) only return total_count without
    # a hasMore/has_more field — compute it from total vs. returned items.
    explicit_has_more = raw.get("hasMore") if raw.get("hasMore") is not None else raw.get("has_more")
    if explicit_has_more is not None:
        has_more = bool(explicit_has_more)
    else:
        has_more = total > 0 and (resolved_offset + len(items)) < total

    return {
        "data": items,
        "pagination": {
            "page": page,
            "limit": resolved_limit,
            "total": total,
            "has_more": has_more,
        },
    }


def normalize_single(raw: Dict[str, Any], key: str, model: Type[T]) -> T:
    """Instantiate *model* from the resource-specific key in *raw*.

    Args:
        raw:   Raw JSON dict returned by the worker.
        key:   Resource-specific key (e.g. "task", "goal", "document").
        model: Pydantic model class to instantiate.
    """
    obj = raw.get(key) or raw.get("data")
    if obj is None:
        raise ValueError(f"Expected key '{key}' in response: {raw!r}")
    return model(**obj)


def normalize_list(raw: Any, key: str, model: Type[T]) -> List[T]:
    """Extract a plain list from a response that may be keyed or bare.

    Args:
        raw:   Raw JSON dict (or list) returned by the worker.
        key:   Resource-specific key if response is a dict.
        model: Pydantic model class to instantiate each item.
    """
    if isinstance(raw, list):
        return [model(**item) for item in raw]
    if isinstance(raw, dict):
        items = raw.get(key) or raw.get("data") or []
        return [model(**item) for item in items]
    return []
